### ISSUE TITLE
KML Support

### ISSUE THREAD

**Issue by andr3nun3s** [CesiumGS/cesium#873]
This issue has been heavily edited since Andre's original write-up and is now maintained by @mramato.

The initial KML implementation was started as part of the 2013 Google Summer of Code and resulted in many core features being implemented by @andre-nunes in his [branch](https://github.com/andre-nunes/cesium/tree/kml).  Current work is taking place in the [kml branch](https://github.com/AnalyticalGraphicsInc/cesium/tree/kml) of the official Cesium repository.  Since KML is a large specification, the plan is to implement what we see as the minimum viable product first and officially release that as part of Cesium.  We will then continue to add additional functionality on a feature by feature basis with each Cesium release.

The below task list tries to capture everything at a fairly high level for each phase.  There are many hidden features and requirements here and listing them all isn't useful.  For example, we are going to need polygon and polyline draping on terrain as a core Cesium feature (#2172).  We will add new tasks when we encounter an issue as we implement each of the high-level types.  If you have a strong need for a feature that we haven't gotten to yet, or feel we're missing something in the below list; please don't hesitate to let us know.
#### Related Links

[Mailing-list topic](https://groups.google.com/d/msg/cesium-dev/xz2xwKogybY/jt8h174rknwJ)
[Google Documentation](https://developers.google.com/kml/documentation/)
[OGC Specification](http://www.opengeospatial.org/standards/kml)
- [x] KmlDataSource
  - [x] KMZ support
  - [x] Proxy support
- [x] [Folder](https://developers.google.com/kml/documentation/kmlreference#folder)
- [ ] [Placemark](https://developers.google.com/kml/documentation/kmlreference#placemark)
  - [x] Geometry
    - [x] [Point](https://developers.google.com/kml/documentation/kmlreference#point)
    - [x] [LineString](https://developers.google.com/kml/documentation/kmlreference#linestring)
    - [x] [LinearRing](https://developers.google.com/kml/documentation/kmlreference#linearring)
    - [x] [Polygon](https://developers.google.com/kml/documentation/kmlreference#polygon)
    - [x] [MultiGeometry](https://developers.google.com/kml/documentation/kmlreference#multigeometry)
    - [x] [gx:MultiTrack](https://developers.google.com/kml/documentation/kmlreference#gxmultitrack)
    - [x] [gx:Track](https://developers.google.com/kml/documentation/kmlreference#gxtrack)
    - [ ] [Model](https://developers.google.com/kml/documentation/kmlreference#model)
- [ ] [Overlay](https://developers.google.com/kml/documentation/kmlreference#overlay)
  - [x] [GroundOverlay](https://developers.google.com/kml/documentation/kmlreference#groundoverlay)
  - [x] [ScreenOverlay](https://developers.google.com/kml/documentation/kmlreference#screenoverlay)
  - [ ] [Super Overlays](https://developers.google.com/kml/documentation/kml_21tutorial#superoverlays)
  - [ ] [PhotoOverlay](https://developers.google.com/kml/documentation/kmlreference#photooverlay)
- [x] [NetworkLink](https://developers.google.com/kml/documentation/kmlreference#networklink)
  - [x] `onExpire`, `onInterval`, `onStop`
  - [ ] `onRegion`, `onChange`, `onRequest`
- [ ] [NetworkLinkControl]
(https://developers.google.com/kml/documentation/kmlreference#networklinkcontrol)
  - [x] Everything except updates
  - [ ] Updates
- [ ] Styles
  - [x] [LineStyle](https://developers.google.com/kml/documentation/kmlreference#linestyle)
  - [x] [PolyStyle](https://developers.google.com/kml/documentation/kmlreference#polystyle)
  - [x] [IconStyle](https://developers.google.com/kml/documentation/kmlreference#iconstyle)
  - [x] [LabelStyle](https://developers.google.com/kml/documentation/kmlreference#labelstyle)
  - [x] [BalloonStyle](https://developers.google.com/kml/documentation/kmlreference#balloonstyle)
  - [x] [styleUrl](https://developers.google.com/kml/documentation/kmlreference#styleurl)
  - [ ] [StyleMap](https://developers.google.com/kml/documentation/kmlreference#stylemap)
    - [x] normal
    - [ ] highlight - https://github.com/CesiumGS/cesium/issues/6059
    - [x] locally defined
  - [ ] [ListStyle](https://developers.google.com/kml/documentation/kmlreference#liststyle) (requires `DataSourceBrowser` for full implementation)
- [x] Additional `<Link>` features
  - [x] `<refreshMode>`
  - [x] `<refreshInterval>`
  - [x] `<viewRefreshMode>`
  - [x] `<viewRefreshTime>`
  - [x] `<viewBoundScale>`
  - [x] `<viewFormat>`
  - [x] `<httpQuery>`
- [ ] Same `<Link>` features for `<Icon>`
- [ ] [LookAt](https://developers.google.com/kml/documentation/kmlreference#lookat)
- [ ] [Camera](https://developers.google.com/kml/documentation/kmlreference#camera)
- [ ] [ExtendedData](https://developers.google.com/kml/documentation/kmlreference#extendeddata)
  - [x] `<Data>`
  - [ ] `<Schema>` and `<SchemaData>`- https://github.com/CesiumGS/cesium/issues/8931
  - [ ] Arbitrary XML data
- [x] Resolve KMZ links inside of an entity description.
- [x] Turn plain text urls in descriptions into links.
- [x] [gx:LatLonQuad](https://developers.google.com/kml/documentation/kmlreference#gxlatlonquad)
  - [ ] Trapezoidal Texture Projection
- [ ] [Region](https://developers.google.com/kml/documentation/kmlreference#region)
- [x] root:// notation for older KML documents
- [ ] [gx:Tour](https://developers.google.com/kml/documentation/kmlreference#gxtour)
- [ ] [gx:drawOrder](https://developers.google.com/kml/documentation/kmlreference#gxdraworder)
- [ ] `<gx:altitudeOffset>` for LinearRing and LineString
- [ ] `<gx:angles>` for Track and MultiTrack (applies to both model and billboard).
- [ ] `<Icon>` gx extensions `x`, `y`, `w`, and `h` for `GroundOverlay`
- [ ] gx extensions for `LineStyle`
- [ ] `xal:AddressDetails`
- [x] `altitudeMode`
  - [x] `absolute`
  - [x] `clampToGround`
  - [ ] `relativeToGround` (doesn't work on terrain).
- [ ] `gx:altitudeMode`
- [x] visibility
- [ ] `ge:Track` gets billboard heading from angles or direction of travel.
- [ ] Support for TIFF images. [forum discussion](https://groups.google.com/d/msg/cesium-dev/_MoqjO-PgcE/pPOZLwvZ15oJ)
- [ ] Password protected kmz files.
- [ ] Password protected network links.
- [ ] Ability to specify default icon when load fails.

**Random notes**
- We need to match handling of missing x/y/w/h values when using Placemark gx sub-region
- All billboards KML produces are currently square, Google Earth instead scales the longer edge to 32 while maintaining aspect ratio.  This can be seen in the KMZ Sandcastle demo, which has squate flags in Cesium and rectangular flags in GE.
- Cesium doesn't have any way to support `BalloonStyle:displayMode`. I'm not sure we care.
- We currently replace BalloonStyle text `$[geDirections]` with nothing.  If we ever provide out of the box routing, we can use that instead.
  (https://github.com/gregjacobs/Autolinker.js) can be used to do this.
- We may be able to use `createObjectURL` for loading data from a zip file (which is currently using data uris). The problem here is that you need to call `revokeObjectURL` in order to clean things up when you're done.
- We need to support passing an ellipsoid to `KmlDataSource` to support other planets.
- We may want to expose some way to customize default values, either by exposing default entity graphics directly, or exposing a limited set of options (like GeoJSONDataSource does)
- Clicking on a link in a balloon description that returns KML has special behavior in GE.  While we probably won't mirror that directly in Cesium, we should provide a mechanism to be able to do so (maybe through events).
- `zip.js` has the ability to report progress which we currently don't utilize. `loadXmlFromZip` and `loadDataUriFromZip` in KmlDataSource.js would be the place to start if we ever want to add it.  We just need to pass a `onprogress` callback as the third parameter to `entry.getData`
- It would be nice to have a KML animation demo that uses TimeSpan, similar to the US states formation demo in Google's own documentation.
- Google Earth silently ignores errors with invalid dates (such as having garbage in a `when` block).  If we want to match Google Earth, we should too.  Rather than try/catching JulianDate.fromIso8601, we should add a exceptionless version that returns undefined.

**Comment by mramato** [comment]
Hey @andre-nunes I know we haven't talked much but I just wanted to let you know that we just opened a PR for KML to go into master (link above this post).  There's still a lot of work to do with KML overall, but it's looking really good for a bunch of use cases.  That's again for getting the ball started on this!

**Comment by andre-nunes** [comment]
Hi @mramato, I'm really glad to see some of my work being finally merged into cesium, let me know if there's something I can help out with in my free time. Oh and I just noticed that you've updated GSoC's ideias for 2015, is Cesium applying this year? I might shoot you an email regarding that ;)

**Comment by mramato** [comment]
Yes, we're applying. I assumed you graduated already. Definitely let me know if you're interested and what you want to work on.

**Comment by GatorScott** [comment]
Some, because I didn't try them all, of the KML files available from http://activefiremaps.fs.fed.us/googleearth.php result in error.

An error occurred while loading the file: conus_latest_modis.kml
An error occurred while loading the file, which may indicate that it is invalid.  A detailed error report is below:

Request has failed.

**Comment by GatorScott** [comment]
The file https://www.google.com/maps/d/kml?mid=zp8nK_5H0MFQ.kzTmU5XK-qJQ results in the following:

An error occurred while loading the file: 2015 California Fire Map.kmz
An error occurred while loading the file, which may indicate that it is invalid.  A detailed error report is below:

TypeError: Cannot read property 'length' of undefined
TypeError: Cannot read property 'length' of undefined
    at Autolinker.HtmlParser.Autolinker.Util.extend.parse (http://localhost:8080/Source/ThirdParty/Autolinker.js:947:24)
    at Autolinker.link (http://localhost:8080/Source/ThirdParty/Autolinker.js:359:15)
    at processDescription (http://localhost:8080/Source/DataSources/KmlDataSource.js:1225:27)
    at processFeature (http://localhost:8080/Source/DataSources/KmlDataSource.js:1288:9)
    at processPlacemark (http://localhost:8080/Source/DataSources/KmlDataSource.js:1331:17)
    at processDocument (http://localhost:8080/Source/DataSources/KmlDataSource.js:1319:21)
    at processFolder (http://localhost:8080/Source/DataSources/KmlDataSource.js:1327:9)
    at processDocument (http://localhost:8080/Source/DataSources/KmlDataSource.js:1319:21)
    at processFeatureNode (http://localhost:8080/Source/DataSources/KmlDataSource.js:1477:13)
    at http://localhost:8080/Source/DataSources/KmlDataSource.js:1502:13

**Comment by mramato** [comment]
Thanks @GatorScott.  The `Request has failed.` is a CORS issue. Just like imagery, any KML files loaded via url to another website or that contain a NetworkLink to another website either needs CORS to be enabled on that server or it has to go through a proxy. I'm writing up a tutorial on using KML in Cesium and I plan on having a section covering this.

That being said, once I used a proxy, I got a lot of errors about invalid styles; so I'll definitely look into and fix the issue for all KML on that page.

The length issue appears to be a clear cut bug caused by KML objects with no description but an empty `ExtendedData` node.  It's an easy fix that I'll have a PR open for in a few minutes.

Thanks again for the link.

**Comment by mramato** [comment]
#2538 fixes the length problem.

**Comment by mramato** [comment]
#2539 fixes the rest of the issues.  All of the files @GatorScott linked to should now load and work in Cesium (though loading directly from the URL, or ones that use NetworkLink still require use of a proxy).

**Comment by GatorScott** [comment]
The "2015 California Fire Map KML" works now but is displaying the placemark names using a serif font that is pretty illegible.

**Comment by mramato** [comment]
> The "2015 California Fire Map KML" works now but is displaying the placemark names using a serif font that is pretty illegible.

It's using the browser's default san-serif font, but I agree it could be done better.  I also want to allow people to select the font to use for labels as well.  Cesium's label rendering as a whole needs improvements, this problem is not specific to KML.  I'll write up an issue, since I don't think we have one right now.

**Comment by GatorScott** [comment]
You wrote "san-serif" when it should be "sans-serif" which is French for without serif.

**Comment by mramato** [comment]
Looks like that typo is also in the KML code, so it's not even using `sans-serif` but instead whatever the default browser font is.  Unfortunately, fixing it doesn't fix the overall problem so we still have a bug somewhere in `LabelCollection` or `writeTextToCanvas`.  Thanks for the heads up.

**Comment by GatorScott** [comment]
Could you make that a bold sans-serif by default?  It looks way better.
![image](https://cloud.githubusercontent.com/assets/4126790/6473341/d37235ac-c1c6-11e4-8b42-59c7ce60cf0a.png)

I tried to get a non-bold picture but the California fire file doesn't work from cesiumjs.org.

**Comment by mramato** [comment]
I opened #2549 which majorly improves font issues (but is going to probably be partially held up by #2130).  We could go with a bold in the meantime, I also recommended a monospace font in a comment in that PR which looks good without bolding.

**Comment by sabrick** [comment]
Is there a way to control z-level for polygons loads through KMLs?

I have placemarks loaded into my viewer:
![image](https://cloud.githubusercontent.com/assets/11263423/6815340/8a57dd9a-d256-11e4-8652-4d6dde89c1bc.png)

But when I load a polygon over the are the placemarks are buried behind it, barely visable:
![image](https://cloud.githubusercontent.com/assets/11263423/6815345/a1dff308-d256-11e4-8e54-1144942db17c.png)

Is there a way to make it so the placemark undeaneath can come to the top? A way to control the z-level? I believe a problem might be that the polygon kml acting at the large blue overlay has the attribute: 
<gx:altitudeMode>clampToSeaFloor</gx:altitudeMode>

And I noticed that "gx:altitudeMode" has yet to incorporated into the API, perhaps that is the problem?

**Comment by mramato** [comment]
I swear I answered this yesterday, but I guess I never submitted the comment.  I just [posted an answer](https://groups.google.com/d/msg/cesium-dev/_nw-F8armtQ/NMiXNRaqmIQJ) to the same question on the mailing list.  I assume that was you too?

**Comment by maburiza** [comment]
Hi, How can I get the coordinates from Kml file using JavaScript

**Comment by Hoverbear** [comment]
Hi folks, is there anyone working on `lookAt` right now? If not, could someone point me at where I might want to look to add this or at least access the data in the KML? Thanks.

**Comment by mramato** [comment]
@maburiza and @Hoverbear in the future, please use [the forum](https://groups.google.com/forum/?fromgroups#!forum/cesium-dev) for general questions and discussions.  More people will see it and you're likely to get a faster answer.

@maburiza We don't have a way to directly get the extent of an entire file (since it changes over time), but you can iterate the loaded entities and discover any information you need.

@Hoverbear No one is working on lookAt, but if you are interested in adding the capability, we would be willing to look at a pull request.  My first approach would be to modify `processFeature` in `KmlDataSource` to load the `lookAt` element and convert it and assign it to the `viewFrom` property on the `Entity` being created.  There might be some roadblocks to this approach but if you take it as far as you can and open a PR, we can help you finish it.

Please create a forum topic to continue any further discussion.

**Comment by maikuru** [comment]
There are some sample KML files for regions in this forum thread
https://groups.google.com/forum/#!topic/cesium-dev/7BTA9RCCdIc

**Comment by tonpo** [comment]
another resources for regions. I find them more clear for me than the KML reference:

1. the standard: 
http://www.opengeospatial.org/standards/kml

2. this tutorial: 
http://www.google.com/earth/outreach/tutorials/region.html

**Comment by 13kenney** [comment]
There appears to be no support for NetworkLink/Link/viewRefreshMode, viewRefreshTime and viewFormat.  Those would be very helpful.  We use:

<viewRefreshMode>onStop</viewRefreshMode>    <viewRefreshTime>0</viewRefreshTime>    <viewFormat>BBOX=[bboxWest],[bboxSouth],[bboxEast],[bboxNorth]</viewFormat>

to have the network link update when the user stops moving, and to limit the view to the bounding box.

**Comment by mramato** [comment]
@13kenney Yes, this is explicitly marked not-support in the above roadmap.  We will be implementing sometime over the next few months (but don't have an exact ETA).

**Comment by 13kenney** [comment]
Thanks, great to hear it is coming!

Mark

**Comment by RameshThangamuthu** [comment]
Hi, 

Anybody working on ScreenOverlay feature? Any ticket in place or plan for it?

Regards,
Ramesh

**Comment by mramato** [comment]
No one is actively working on it, but it's definitely on the roadmap.  We are hoping to do another round of KML improvements early next year, but I don't have any kind of ETA for you.

**Comment by RameshThangamuthu** [comment]
Thanks for the quick response. (Until then, we will parse the KML document and use HTML/JS features to overlay the image on Cesium maps.)

**Comment by mengma16** [comment]
Hi
Anybody working on Placemark Geometry-Model feature? Any plan for it? Our project used to display 3D models created by google-skechup on google-earth. Now we are seeking a solution to display these models on the map.

**Comment by vanurin** [comment]
Hello,
It seems to me that Cesium interprets KML absolute `<altitudeMode>` as altitude from ellipsoide, while in KML [specs ](https://developers.google.com/kml/documentation/altitudemode)it should be an altitude from sea level (or EGM96 if my guess about Google Earth sea level is true).
Please, tell if there is any work towards altitudeMode support?

**Comment by mramato** [comment]
> Please, tell if there is any work towards altitudeMode support?

It is on our roadmap, but there's no ETA.

**Comment by vanurin** [comment]
> It is on our roadmap, but there's no ETA.

Thank you for response! And good luck with your roadmap!

**Comment by tfili** [comment]
@mramato Here is what should be checked off

For NetworkLinks

In `<Link>` all these are done
```
 <refreshMode>
 <refreshInterval>
 <viewRefreshMode> (onStop only)
 <viewRefreshTime>
 <viewBoundScale>
 <viewFormat>
 <httpQuery>
```

For `<Icon>` there isn't any refreshing but these now work for the first load
```
 <viewBoundScale>
 <viewFormat>
 <httpQuery>
```

`<NetworkLinkControl>` has support for
```
<cookie>
<expires>
<maxRefreshInterval>
```

These are also done
 - StyleMap - locally defined
 - root:// notation for older KML documents

**Comment by 13kenney** [comment]
Thank you, wonderful news about the Network Links support!

**Comment by zengtaouou** [comment]
Cesium KML not support 3d model?

**Comment by hpinkos** [comment]
NetworkLinkControl support requested in #4404

**Comment by micelic** [comment]
I'm porting an application from GE to Cesium; it relies on NetworkLinks updates heavily, a Is there any roadmap for such a feature? If the answer is No.. does Cesium have Network  Links but not using KML files but some other format I may have to convert to ? Thank You

**Comment by hpinkos** [comment]
Hi @micelic!  Cesium does have support for network links, but performance may be slow depending on how your KML is organized.  If performance is an issue, I would recommend converting to CZML.  Other users that have migrated from GE->Cesium have seen significantly better performance with CZML than KML.  You can learn about the format here: https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/CZML-Guide

In the future, please post questions like this on our forum: https://groups.google.com/forum/?hl=en#!forum/cesium-dev
There is higher visibility there so it is more likely that others will see your question and be able to help answer it.  And it may help out other members of the community who have a similar question.   Thanks!

**Comment by micelic** [comment]
Thank You.

I would like to try the network links first to see the performance hit.. If it is an issue I will think to convert to CZML.

Is there anything documented on Cesium and network links for KML.

The documentation on this topic is limited.

I appreciate your help.

**Comment by prieks** [comment]
Any plans for Screenoverlays? Previous comments have indicated it was on roadmap, but not sure if this work has started?

**Comment by hpinkos** [comment]
@prieks we will likely not add screen overlay support because you can easily use HTML and CSS to create and position elements on top of the viewer.  You can see this example for how to do so programmatically: http://analyticalgraphicsinc.github.io/cesium-google-earth-examples/examples/screenOverlay.html

**Comment by pjcozzi** [comment]
Although Cesium may not have a general screen overlay API, I'm pretty sure that we would add ScreenOverlay support for KML if enough users need it or someone is able to contribute it.

**Comment by florentdu** [comment]
Any plans for <gx:angles> ? It seems to me that display an aircraft orientation based on pitch, roll, heading is a nightmare in Cesium. KML <gx:angles> is not implemented. CZML orientation must  be computed in fixed Earth reference. Even in javascript, pitch, roll, heading must me transformed with heading-pitch-roll to Quaternion and Transforms.eastNorthUpToFixedFrame.

**Comment by hpinkos** [comment]
Hello @florentdu,  gx:angles is on our list but I don't think we'll have a chance to work on it in the near future.  We would gladly review a pull request though if you have some time to work on implementing this!  You would need to make changes to [KmlDataSource](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/DataSources/KmlDataSource.js).  You can find information on checking out and building the code base in our [Build Guide](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/BuildGuide) and see our [Contributing Documentation](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#opening-a-pull-request) for instructions on opening a pull request. Thanks!

**Comment by burn123** [comment]
When will there be support for infoboxes using KML documents? Currently, I can only get the infobox to show up for my KML file if I wrap the `<Document>` in a `<Folder>`

**Comment by mramato** [comment]
@burn123 I'm not sure what you mean.  Infoboxes are already supported for KML.  If you have a KML file or snippet that you feel should work, but doesn't, please share it and we'll take a look.

**Comment by burn123** [comment]
In my app, the infobox won't appear if I don't have the Document wrapped by a Folder element. So for example, in this snippet I can't get the Placemark description to show up in the infobox, because the infobox never appears

    <?xml version='1.0' encoding='UTF-8'?>
    <kml xmlns="http://www.opengis.net/kml/2.2" xmlns:gx="http://www.google.com/kml/ext/2.2" xmlns:kml="http://www.opengis.net/kml/2.2" xmlns:atom="http://www.w3.org/2005/Atom">
	<Document>
            <name>Beam Footprint Plot</name>
            <open>1</open>
            <Placemark>
			<name>Beam 51</name>
			<description><a href="http://google.com">SMAC instance id: 12</a><br /> Access Network id: 111 <br /> SMAC domain id: 324 <br />SMAC base station id: 01:03:24:68:00:00 <br />Logical beam id: 51 <br />Physical beam id: 307 <br /> CSA id: 551</description>
            </Placemark>
        </Document>
    </kml>

But if I simply wrap it in a folder tag, it works in my app

    <?xml version='1.0' encoding='UTF-8'?>
    <kml xmlns="http://www.opengis.net/kml/2.2" xmlns:gx="http://www.google.com/kml/ext/2.2" xmlns:kml="http://www.opengis.net/kml/2.2" xmlns:atom="http://www.w3.org/2005/Atom">
      <Folder>
	<name>Temporary Places</name>
	<open>1</open>
	<Document>
             ......
        </Document>
      </Folder>
    </kml>

In the official Cesium viewer, both work so I'm not sure what makes my app unable to render the infobox for the first example

**Comment by thw0rted** [comment]
Is there anything on the roadmap for how to support large KML data sets?  I've seen some discussion about plans for `Region` support, eventually, and that's good if you have control over the source, but I'd also like to allow my users to bring their own KML and handle that as gracefully as possible.

Right now, when you load a large data set (just Googling around I found [this 15MB example](http://rgis.unm.edu/gstore/datasets/0660a796-c062-4d37-81d5-070697147acb/gniswells.derived.kml) ), nothing seems to happen for some time, then the render loop crashes out with "Array buffer allocation failed".  (I tried to make a Sandcastle example to show this, but ran into CORS issues -- sorry!)

I'd like to see better error handling, certainly, but also maybe we could have the ability to limit how many entities are loaded from the file?  Obviously splitting large sets into Regions (once supported), or parsing server-side and exposing WMS / WFS / 3D Tiles / whatever would be best in the long run, but until we can provide that I'm stuck handling KML client-side.

**Comment by hpinkos** [comment]
Hello @thw0rted!  Our strategy for handling large KML datasets will be to process them into 3D tiles server-side.  This is one of the processing tools we plan to make available on [cesium.com](https://www.cesium.com/).  We do want to improve loading and error handling to prevent large datasets from freezing and crashing, but I don't think there's much we have planned to improve rendering support client side.  It's difficult because everything has to be stored in memory and it's just too much data for the browser to handle all at once.

**Comment by thw0rted** [comment]
I don't think it's worth a whole issue, but I wanted to point out that the docs for KMLDataSource say that the static `.load` method's `options` paramter is optional -- this is not true.  The static `load` is a thin wrapper that makes a new instance and calls its prototype `.load`, but it won't get past the constructor if the `options` aren't specified, since `camera` and `canvas` are required.

**Comment by hpinkos** [comment]
Thanks @thw0rted!  I fixed the doc here: #5961

**Comment by thw0rted** [comment]
Uh, am I maybe the only one actually using the KML capability right now?  I sort of assumed it would have lots of users but this thread has been suspiciously quiet :D

I am back because I wanted to ask about Google Earth compatibility.  We have users on Google Earth Pro 7.1 and 7.3, and both have an infuriating feature -- the default Placemark icon set uses links to `http://maps.google.com` (hard-coded protocol and all :( ), for example `http://maps.google.com/mapfiles/kml/shapes/placemark_circle.png`.  That would be fine, except that our network is not on the internet... but Google Earth Pro still happily fills in the correct image!  They must have an internal pre-fetched representation of the default icons and silently use those instead of looking for the actual file online.

What does this have to do with Cesium?  Well, there are a number of compatibility / interoperability features on the roadmap that specifically cater to the Google "flavor" of KML.  Ideally, this "feature" (ugh) of Google Earth could also be replicated.  More generally, maybe we could pass a function (or Proxy?) to the KMLDataSource that is allowed to rewrite resource URIs before they're requested from the network?  This would be more work -- we'd have to grab and host copies of the default icon set somewhere -- but it would also be helpful in other situations, like some legacy KML sources we found where the root file is available over HTTPS (good!) but it includes NetworkLinks over HTTP (bad!).

Is something like that possible today?  If not, could it be added to the roadmap?  Should I file a separate issue?

**Comment by hpinkos** [comment]
@thw0rted yes, can you please write up a separate issue?  That will make it easier to discuss since it's a specific case. Thanks!

**Comment by ggetz** [comment]
`onRegion` networkLink event support requested on the forum: https://groups.google.com/forum/#!topic/cesium-dev/MOuIjyW4pew

**Comment by bbehling-trimble** [comment]
Has terrain support been added? When I add a KML and a terrain provider, the KML moves around when I pan the globe. When I turn off terrain, its fine. Also tried setting clampToGroup: true and still an issue.

**Comment by hpinkos** [comment]
@bbehling-trimble yes, some terrain support has been added.  In addition to passing in the `clampToGround: true` flag to the `KmlDataSource`, you need to make sure your KML has the correct altitude mode set for placemarks. ie `<altitudeMode>clampToGround</altitudeMode>`.  Note that altitude modes like `clampToSeaFloor` are not supported, and `relativeToGround` has support for some placemark types but not all.  If the altitude mode is not supported there should be a warning logged to the console.

**Comment by mramato** [comment]
Thanks for the example file.  At first glance, this looks like a bug.  I'll take a closer look and let you know.

**Comment by mramato** [comment]
@bbehling-trimble KML requires `<tessellate>1</tessellate>` be set on line strings in order for them to conform to terrain.  None of the strings in the file you posted have this set.  I updated the file and now it loads as expected in Cesium.

[Dimensions2018OffsiteScans2.zip](https://github.com/AnalyticalGraphicsInc/cesium/files/2501430/Dimensions2018OffsiteScans2.zip)

That being said, I'm not sure why the original file was loading on terrain in Google Earth.  Longer lines I tested in Google Earth do not follow the terrain unless tessellate is set.  This is probably another case of Google not following their own format and having undocumented behavior (happens all of the time).

I'll look more into it and possibly open a PR to better match GE (if I can figure out the "rules") but for now the most correct thing to do is add tessellate to the LineString.

**Comment by bbehling-trimble** [comment]
@mramato Thanks!

**Comment by thw0rted** [comment]
This thread has been quiet again for a while!  I'd like to make another quick suggestion -- again, it can be its own issue if need be, but it's very small.  I'm using the KmlDataSource in a context where it would be helpful to know if the loaded file has NetworkLinks or not.  I'd like to show the user whether or not they can expect the data to periodically update itself.  I don't need tons of details, but even something like a `hasNetworkLinks` boolean (implemented as `get() { return this._networkLinks.length > 0; }`) would be useful as part of the public API so I don't have to dig around under the hood.

**Comment by OmarShehata** [comment]
I just got a request for supporting `SchemaData` over the support email. Is this something that would be easy to support?

**Comment by gogo1211** [comment]
Hello, @mramato 
https://github.com/AnalyticalGraphicsInc/cesium/issues/873#issuecomment-76861511
Please let me know how to handle that CORS issue while loading KML. #8381

**Comment by OmarShehata** [comment]
@gogo1211 Hannah's response here https://github.com/AnalyticalGraphicsInc/cesium/issues/8381#issuecomment-552457568 explains it best. You need to configure this on your server host. If you need additional help please open a [thread on the Cesium forum](https://groups.google.com/forum/#!forum/cesium-dev).

**Comment by gogo1211** [comment]
@OmarShehata , thank you!

**Comment by paullenhardt** [comment]
I have come across an odd issue relating to kml 'Folder' support (which is currently marked as supported).  If I load a kml file with folders, none of the data shows within cesium and I get no errors on the console.

``` javascript
var viewer = new Cesium.Viewer('cesiumContainer');
var options = {
   camera: viewer.scene.camera,
   canvas: viewer.scene.canvas
};
viewer.dataSources.add(Cesium.KmlDataSource.load('test_wind.kml', options)).then(function(data){ 
   console.log("Data loaded:", data);
});
```
The above outputs "Data loaded" and displays nothing within cesium.

As soon as I strip the folder tags out of kml, everything seems to work.
``` javascript
var viewer = new Cesium.Viewer('cesiumContainer');
var options = {
   camera: viewer.scene.camera,
   canvas: viewer.scene.canvas
};
viewer.dataSources.add(Cesium.KmlDataSource.load('test_wind_no_folders.kml', options)).then(function(data){ 
   console.log("Data loaded:", data);
});
```

test_wind.kml
``` xml
<?xml version="1.0" encoding="UTF-8"?>
<kml xmlns="http://www.opengis.net/kml/2.2">
<Document>
  <Style id="deg180speed50">
    <IconStyle>
      <scale>1.125</scale>
      <heading>180</heading>
      <Icon>
        <href>track-0.png</href>
      </Icon>
    </IconStyle>
  </Style>
  <Folder id="time1499886000.0">
    <name>1499886000.0</name>
    <visibility>0</visibility>
    <Folder id="altitude24000">
      <name>24000</name>
      <visibility>0</visibility>
      <Placemark>
         <styleUrl>#deg180speed50</styleUrl>
         <Point>
            <altitudeMode>clampToGround</altitudeMode>
            <coordinates>-98,38</coordinates>
         </Point>
      </Placemark>
    </Folder>
  </Folder>
</Document>
</kml>
```

test_wind_no_folders.kml
``` xml
<?xml version="1.0" encoding="UTF-8"?>
<kml xmlns="http://www.opengis.net/kml/2.2">
   <Document>
      <Style id="deg180speed50">
         <IconStyle>
            <scale>1.125</scale>
            <heading>180</heading>
            <Icon>
               <href>track-0.png</href>
            </Icon>
         </IconStyle>
      </Style>
      <Placemark>
         <styleUrl>#deg180speed50</styleUrl>
         <Point>
            <altitudeMode>clampToGround</altitudeMode>
            <coordinates>-98,38</coordinates>
         </Point>
      </Placemark>
   </Document>
</kml>
```

**Comment by thw0rted** [comment]
I haven't tried loading your sample KML but you have both of your (nested?) folders' `<visibility>` tags set to 0, i.e. hidden.  I spent 10 seconds searching the source for `KMLDataSource` and it [looks like that tag should be respected](https://github.com/AnalyticalGraphicsInc/cesium/blob/1.63.1/Source/DataSources/KmlDataSource.js#L1759) so setting it to 1 is likely to fix your issue.

**Comment by paullenhardt** [comment]
Well that would make plenty of sense.  Apparently you can stare at an issue for hours and still never see the obvious.  

GoogleEarth seems to ignore the `<visibility>` tag (I originally had set the `<visibility>` to '0' within my kml generator to try to suppress the display upon KML loading and forgot about removing the `<visibility>` tag when it had no effect on GoogleEarth) so I am glad to see that Cesium honors that tag.

The nested folders was simply because I pulled the example from a much larger data-set where there was more data which the nesting provided organization to.

Thanks for the assist in debugging despite the obvious blunder on my part.

**Comment by ainthek** [comment]
gxlTrack when rendered on terrain shows only icon, and path is underground. 
Minimal KMLs are attached in sandcastle demos.

Let's have 4 points and 2 flavors of KML, one as LineString and one as gx:Track

First example: LineString works (after some suggestions on forum)  https://sandcastle.cesium.com/index.html#c=bVRtT9swEP4rVr80lTKn6XipSgjTQOzDQCC12j4skzDOtbVw7Mh2Wgriv++cN1pYVTW95557vwvXyjqyEbAFQ86Jgi25BCuqgv6qsWDIa/FSK8eEAjMMyWumCHFgDAL3Rm9EDmbWmXEDzMFvbWS+aCjBKMzU2+gsUxtmiC6dwJgY65V4P4RwVoBhZNZmQS0HBbRBw56DxvYzx6NhS5GsKBf6h9GVymfEmQow7Blpwr5oXSyYWYHrkKdCYhIPycUz/tmAsZjWeTaI6TgbEFBc50KtEKjc8ss0G1ykmUq8DdKVRXztXDmLou12S3UJaiUsVeAipEQTOkEfNXG2ej7krrReSUxdFzUVnl1DT30Vyb1kHApmntKmqERhG9LbHbnB3idRLbWaHCw3ou5mulgLS/Dr1kCk5syDRC9JscPfpeBAk2if37rwTufOYJ0tgpgDa0FKnGEaJ9Ge1DOYdMJVOdzqHNKDrifRga634Fob7CZ6sWl8SuPJeHI8DY+mdBKPj04npMHGxw0Un8YnLRR/Peow1DefuNVNev7J9DiJ9mP0gaMPFSbRfoMTPwF8PrSrWTJj+yu4uru9r+Wg29zS6EJYQH276j8LecUcm+vKYIelZnnQuKD14xr5TewA4+AmD5yfNq5FNhiF3SGM/D62DrdrUEEbJiTLSnFPCfI+yKg/mvYO3lWWsjzfp561zPfNx8zf9RQUjkqg3YbJCuyf8V9v8VbbZWrOVM6Zdbiq6HehtXxk5nvlHOYzvJY74jTpezncS/ZTiku5W+iuqtFZFFHny0Tiefp/Tp3CW0iGrgk7RGAQDhLrdrLbqm+iKLVxpDIyoBT3tCj9ltroseJP4Ci31s8tiTqjJBcbInI8xg8vNLxUXGLrT3pZSTkXL4DXiPciNgdmfr44yzt8VUi285R1nN40IKV4YCh+tmpL2PP4Dw

Second Example: 
gx:Track, "banner" is OK 
but "track line" is UNDERGROUND
 https://sandcastle.cesium.com/index.html#c=jVX7b9s2EP5Xbv7FNuBQjzzhKOq2Fu2GNU1QCxuwaUAZ6WwToUiDpOy4Rf73HfWKs3RYBMIy777v7vj4ToVW1sFW4A4NXIHCHbxFK+qK/d7YJuOimb7VynGh0Ixn8C1XAEEA2S+/LsCP69ubz9lPnzL4AbiUYBErCwa5RFeC0+DQGCLDHYqVAk3s5vFhOtet0VtRopn32QuiO/xDG1lmLWQyneXqcXqZqy03oDdOUOlU8rcmDkDBKzQc5t1imC1QIWutswFDZPsS462zDiJ5tcn0B6NrVc7BmRop7SW0ab9qXWXcrND1lvtKUhFfkjcP9GeLxlJZV/koYmE+AlSFLoVakaF2y6OLfPQmzVXiOQRXluxr5zbzINjtdkxvUK2EZQpdQJAgZjHFaIDz1cNz7ErrlaTSddVA8cG18NSvIrmVvMCKm/u0XVSiaBvS6z18pCNMgmbWeUq0hRHNbqbZWlig4dYIUhfcG0EvodrT71IUyJLgEN+FWD3MM8OLPhlZHFqLUtIJplESHMwGBJdOuLrEa11i+mzPk+CZb2Ds1qjSOIzOjsLjo+Mwi47n0SmNP5Ogcf0f8PyVwDh8DTAOvweknSi0NmUanbMoDuPTCzi5YHEUnpzHSTB4/wsfnrbw6Dw6ewU8Oj7p8cRtn+gVvHjIc3Zx+gLfGJ7OMwkOL1Pibxu9v3Qy3HBjh8bx7ub6tplPepVujK6ERfJ3sv6tku+44wtdG7pNUvNy0oZgzes94RfOkGImlIdUO3L+ZpME8tF01ot+6rXXBfT7P+nSzGBZq8JDJuWQZDo0iE7zTy7LeFkeQi875JPKqfInP0NFF1MQb8tljfav8G/PeGx4uVpwVRbcOpIlxc20lnfc/Fw7R/WM38u974TDXo4Pin1R4lLuM92vasqcXyOhrtLvA5r8jzMYuzbnmAyj2Sixbi97Af0oqo02DmojJ4yRJKuNF6QN7uriHh0rrPWHlgQ9KSnFFkRJXedfHwBqSaRX63vXspZyIb4itR1qDGL7jOYPlw7yhnqi5HsPWUfpx9bIGHUSmr5kdUs4iPgP

Third Example: 
When no terraiin is specified, also gx:Track works
  https://sandcastle.cesium.com/index.html#c=jZX7b9s2EMf/lZt/sQ041CNPOIq6rUW7YU0T1MIGbBpQRjrbRCjSICk7bpH/fUe94iwdFkOwzOPnHnx8z4VW1sFW4A4NXIHCHbxFK+qK/d7YJuOiGb7VynGh0Ixn8C1XAEEA2S+/LsA/17c3n7OfPmXwA3ApwSJWFgxyia4Ep8GhMeQMdyhWCjR5N582TDd5a/RWlGjmff6CAjj8QxtZZi0ymc5y9Ti9zNWWG9AbJ6h4KvpbEwmg4BUaDvNuOcwWqJC11tnAkLN9yXjrrEMkrzaZ/mB0rco5OFMjpb2ENu1XrauMmxW63nJfSSriS/LmgX5s0Vgq6yofRSzMR4Cq0KVQKzLUbnl0kY/epLlKvA/hypJ97dxmHgS73Y7pDaqVsEyhCwgJYhZTjAacrx6esyutV5JK11WD4oNr8dSvIrmVvMCKm/u0XVSiaBvS6z18pENMgmbUzZRoCyOa3UyztbBAj1sjSF1wbwS9hGpP30tRIEuCQ74LsXqYZ4YXfTKyOLQWpaQTTKMkOBgNBJdOuLrEa11i+mzPk+DZ3OCxW6NK4zA6OwqPj47DLDqeR6f0/JkEzdT/geevBOPwNWAcfg+knSi0NmUanbMoDuPTCzi5YHEUnpzHSTDM/hcfnrZ4dB6dvQKPjk96nnzbT/QKv3jIc3Zx+oJvDE/nmQSHlynxt43eXzoZbrixQ+t4d3N924wnvUo3RlfCIs13sv6tku+44wtdG7pNUvNy0oZgzes98QtnSDETykOqHTl/s0kC+Wg660U/9drrAvr9n3RpZrCsVeGRSTkkmQ4NotP805RlvCwP0cuOfFI5Vf40z1DRxRTkt+WyRvtX+Lf3eGz8crXgqiy4dSRLiptpLe+4+bl2juoZv5d73wuHvRwfFPuixKXcZ7pf1ZQ5v0airtLvA03+xxmMXZtzTIbRbJRYt5e9gH4U1UYbB7WRE8ZIktXGC9IGd3Vxj44V1vpDS4LeKSnFFkRJXedffwHUkkiv1veuZS3lQnxFajvUGMT2mZs/XDrIG+qJku89so7Sj62RMeokNHzp1S3hIOI/

How to fix/workaround samples to make gx:Track and their respecitive lines to work with terrainProvider ON ?

Are there any tests simulating behavior of KML imported data on terrain ? I have studied briefly DataSources/KmlDataSource test case, but is does not seem to cover terrainSpecific rendering which seems to be a problem.

Thanx a lot

**Comment by thw0rted** [comment]
I'm not an expert but it looks like `gx:track` translates to [PathGraphics](https://cesium.com/docs/cesiumjs-ref-doc/PathGraphics.html) under the hood.  [Polylines](https://cesium.com/docs/cesiumjs-ref-doc/PolylineGraphics.html) support `clampToGround` but Path does not have that option.  If you think about it, this makes sense, because a Polyline has its own set of positions independent of the parent Entity, but Path relies on the time-variant `Entity#position`.  The docs for [KMLDataSource#load](https://cesium.com/docs/cesiumjs-ref-doc/KmlDataSource.html) say that it clamps "geometry features (Polygons, LineStrings and LinearRings)" but not the entities themselves.

I've run into similar problems before but I am not loading from KML, so I don't know if my solution would help you.  I compute a `Cartographic` for each position, then pass that to `sampleTerrainMostDetailed`, then if that fails I call `sampleTerrain` with a fixed level -- I picked 11, can't remember why.  Once my Cartographics have been "clamped" then I update the original entity with new position(s).  It's an asynchronous process but at the end the Entity is where I want it to be.

With KML, you'd have to "fix" all the Entities created after `load` finishes.  The problem is that once a [SampledPositionProperty](https://cesium.com/docs/cesiumjs-ref-doc/SampledPositionProperty.html) has been created, there is no public API for retrieving the positions, so you have to poke around in the private part of the class.

It might be worth putting in a new ticket, either to have the `KML#load` option for `clampToGround` also clamp Entity positions when they're created, or to expose the samples in a `SampledPositionProperty` via a getter so that you can fix the positions yourself after the fact.

**Comment by thw0rted** [comment]
Anyone watching this issue might be interested in https://github.com/CesiumGS/cesium/pull/10260 which I just filed.  Not a huge change, but could be useful to some.

**Comment by icarter09** [comment]
Couple of questions on this ticket...

1. What's the status of "StyleMap"->"Highlight" ?
2. If "Highlight" is still a WIP, I would like to help out. What do I need to know about or where should I begin to look for getting Stylemap Highlight implemented? 

Thanks.

**Comment by ggetz** [comment]
Hi @icarter09,

1. I believe StyleMap -> Highlight is still yet to be implemented.
2. It would be great if you could contribute this! I'll look into getting you some more information, but in the meantime [this comment](https://github.com/CesiumGS/cesium/issues/2770#issuecomment-155528406) mentions the areas of the code to look.

**Comment by martimpassos** [comment]
hello, are there plans to support PhotoOverlays? I would like to visualize geolocated images against Google's 3D Tiles

**Comment by selimsezr** [comment]
![image](https://github.com/user-attachments/assets/d2d69f18-f403-4fc3-bed1-5b056b0c1f46)


Hello everyone, I am currently using cesium 1.121 and I loaded the kml file in the picture. It does not say gx:AltitudeMode on the polygon on the ahead, but it does on the polygons on the behind. The code section is below. I want to determine it with cesium settings instead of the value written in the kml. How can I follow?

```
const kmlDataSource = await CesiumJs.KmlDataSource.load(kmlText, {
            camera: CesiumViewer.scene.camera,
            canvas: CesiumViewer.scene.canvas,
            clampToGround: true,
          });
          CesiumViewer.dataSources.add(kmlDataSource);
          const entities = kmlDataSource.entities.values;
  
          entities.forEach((entity:any) => {
            // Poligonların zeminle hizalanması
           if (entity.polygon) {
             entity.polygon.heightReference = CesiumJs.HeightReference.NONE;
             entity.polygon.extrudedHeightReference = CesiumJs.HeightReference.NONE;
           }
```

**Comment by ggetz** [comment]
>  are there plans to support PhotoOverlays? I would like to visualize geolocated images against Google's 3D Tiles

@martimpassos There hasn't been any activity on PhotoOverlays recently. 

If you'd like to add more information about your use case, that may help us prioritize. If you are interesting in contributing, please let us know and we'd be happy to discuss implementation or review a PR. Thanks!

> Hello everyone, I am currently using cesium 1.121 and I loaded the kml file in the picture. It does not say gx:AltitudeMode on the polygon on the ahead, but it does on the polygons on the behind. The code section is below. I want to determine it with cesium settings instead of the value written in the kml. How can I follow?

@selimsezr CesiumJS is a big project, so we use GitHub for feature requests and bug tracking exclusively. Could you please take this question to the [Cesium Forum](https://community.cesium.com/)? On the forum, there are members of the community and developers from the team who can help. Thanks!

**Comment by selimsezr** [comment]
@ggetz While doing research, I found [this](https://community.cesium.com/t/kml-clampping-to-3dtiles/7160/9) topic as a similarity and added it, but I can also open it as a different topic.
